pkg/steps/cluster_claim.go

```diff
@@ -205,14 +201,69 @@ func releaseCluster(ctx context.Context, hiveClient ctrlruntimeclient.Client, cl
 	return nil
 }
 
-func ClusterClaimStep(as string, clusterClaim *api.ClusterClaim, hiveClient ctrlruntimeclient.Client, client loggingclient.LoggingClient, jobSpec *api.JobSpec, wrapped api.Step) api.Step {
+func ClusterClaimStep(as string, clusterClaim *api.ClusterClaim, hiveClient watchingclient.Client, client loggingclient.LoggingClient, jobSpec *api.JobSpec, wrapped api.Step) api.Step {
 	ret := clusterClaimStep{
 		as:           as,
 		clusterClaim: clusterClaim,
 		hiveClient:   hiveClient,
 		client:       client,
 		jobSpec:      jobSpec,
 		wrapped:      wrapped,
+		waitForClaim: func(client watchingclient.Client, ns, name string, timeout time.Duration) error {
```

Comment:
In every case that we write code that looks identical in two places we need to ask if it's attempting to do the same thing, and, if so, how we can make it share the core structure. This code is identical to that which David (Alvaro?) wrote for containers. Consider something like the following:

```go

type evaluator func(obj runtime.Object) (bool, error)

// waitForConditionOnObject uses a watch to wait for a condition to be true on an object.
// When the condition is satisfied or the timeout expires, the object is returned along
// with any errors encountered.
func waitForConditionOnObject(ctx context.Context, client watchingclient.Client, identifier ctrlruntimeclient.ObjectKey, list *ctrlruntimeclient.ObjectList, into ctrlruntimeclient.Object, evaluate evaluator, timeout time.Duration) error {
	lw := &cache.ListWatch{
		ListFunc: func(options metav1.ListOptions) (object runtime.Object, e error) {
			options.FieldSelector = fields.OneTermEqualSelector("metadata.name", identifier.Name).String()
			res := *list
			return res, client.List(ctx, res, &ctrlruntimeclient.ListOptions{Namespace: identifier.Namespace, Raw: &options})
		},
		WatchFunc: func(options metav1.ListOptions) (i watch.Interface, e error) {
			opts := &ctrlruntimeclient.ListOptions{
				Namespace:     identifier.Namespace,
				FieldSelector: fields.OneTermEqualSelector("metadata.name", identifier.Name),
				Raw:           &options,
			}
			res := *list
			return client.Watch(ctx, res, opts)
		},
	}

	// pods in this call here are always expected to exist in advance
	var podExistsPrecondition toolswatch.PreconditionFunc

	waitForContainerStatus := func(event watch.Event) (bool, error) {
		if event.Type == watch.Deleted {
			return false, fmt.Errorf("%s was deleted", identifier.String())
		}
		// the outer library will handle errors, we have no data to review in this case
		if event.Type != watch.Added && event.Type != watch.Modified {
			return false, nil
		}
		return evaluate(event.Object)
	}

	waitTimeout, cancel := toolswatch.ContextWithOptionalTimeout(ctx, timeout)
	defer cancel()
	res := *list
	_, syncErr := toolswatch.UntilWithSync(waitTimeout, lw, res, podExistsPrecondition, waitForContainerStatus)

	// We need to make sure this ends up in the logging client
	if err := client.Get(ctx, identifier, into); err != nil {
		logrus.WithError(err).Warn("failed to get object after finishing watch")
	}
	return syncErr
}
```

Reply:
After 2 abstraction that you have given me, I hope I can learn the level of abstraction.